Implement preview command

mishig25 opened the pull request:
# Dev command with hot reloading

`doc-builder preview datasets ~/Desktop/datasets/docs/source`

I think this will make the experience writing docs better 🙂

https://user-images.githubusercontent.com/11827707/158227378-1bb04d04-4010-48b3-b039-9d2bdc72c4cf.mov

mishig25 commented on src/doc_builder/build_doc.py:
```diff
@@ -213,7 +221,7 @@ def build_mdx_files(package, doc_folder, output_dir, page_info):
         )
 
-    return anchor_mapping
+    return anchor_mapping, src_py_path_mapping
```
`src_py_path_mapping` is used to know wich `mdx` file to updated when a docstring is updated inside `python` file

sgugger commented on src/doc_builder/commands/dev.py:
```diff
@@ -0,0 +1,218 @@
+from doc_builder.commands.build import check_node_is_available, locate_kit_folder
+from doc_builder.commands.convert_doc_file import find_root_git
+from doc_builder.utils import read_doc_config
+from watchdog.events import FileSystemEventHandler
+from watchdog.observers import Observer
```
This is a new dependency, so would prefer to have this as a soft dependency. This means those imports and the class below should be moved under an `if is_watchdog_available:` test.

For the value of `is_watchdog_available` we should use a `try / except` block to see if we can import this or not. Then the `dev_command` would start with a test raising an `ImportError` if `is_watchdog_available` is `False`.

This would solve the doc building issue in the CI.

mishig25 replied:
handled in https://github.com/huggingface/doc-builder/pull/139/commits/317030a82700f2243bf3d55eb744ca4c8b632419

sgugger commented on src/doc_builder/autodoc.py:
```diff
@@ -278,6 +278,19 @@ def get_source_link(obj, page_info):
+    """
+    Find a path to file in which given object was defined.
+
+    Args:
+    - **object_name** (`str`) -- The name of the object to retrieve.
+    -- **package** (`types.ModuleType`) -- The package to look into.
+    """
```
```suggestion
    Args:
    - object_name (`str`): The name of the object to retrieve.
    - package (`types.ModuleType`): The package to look into.
    """
```

sgugger commented on src/doc_builder/build_doc.py:
```diff
@@ -153,6 +156,7 @@ def build_mdx_files(package, doc_folder, output_dir, page_info):
     output_dir = Path(output_dir)
     os.makedirs(output_dir, exist_ok=True)
     anchor_mapping = {}
+    src_py_path_mapping = {}
```
```suggestion
    source_files_mapping = {}
```
Same length but more understandable.

mishig25 replied:
handled in https://github.com/huggingface/doc-builder/pull/139/commits/31c4207a7069d765dc2b633d33d698e1e9da986b

sgugger reviewed (commented):
Looks like a great addition, thanks for working on this! I have a few comments on the variable names and we need to handle `watchdog` as a soft dependency (that the user should install only if they want to use `doc-builder dev`).

The name `dev` is not super intuitive to me, is it something that comes from node? I'd prefer something like `doc-builder preview` (or if you have another idea I'm all ears!).

mishig25 commented on src/doc_builder/commands/dev.py:
```diff
@@ -0,0 +1,218 @@
+        env["DOCS_LIBRARY"] = args.library_name
+        env["DOCS_VERSION"] = args.version
+        env["DOCS_LANGUAGE"] = args.language
+        Thread(target=start_sveltekit_dev, args=(tmp_dir, env, args)).start()
```
this thread should be running as long as the program is running

mishig25 commented:
@sgugger thanks a lot for the suggestions!
1. Renamed command `dev` -> `preview`
2. Added soft dependency `watchdog`
3. Renamed variables as suggested

sgugger commented on src/doc_builder/commands/preview.py:
```diff
@@ -0,0 +1,225 @@
+
+
+def preview_command(args):
+    if is_watchdog_available():
```
To avoid everything being indented here, I'd test the negation and raise the `ImportError`, then just unindent the whole function.

mishig25 replied:
https://github.com/huggingface/doc-builder/pull/139/commits/979e8d981a887d911374d9d5e94fdf2b1e22a0ad

sgugger reviewed (approved):
Thanks for adding all of that! I just left one last styling suggestion.

For the `Thread`, are we sure that exiting will close it (it's not the case with `Process` if I remember correctly, so just want to be 100% sure)

sgugger commented on src/doc_builder/commands/preview.py:
```diff
@@ -175,9 +175,9 @@ def preview_command(args):
             # files/folders cannot have a name that starts with `__` since it is a reserved Sveltekit keyword
             for dir in output_path.glob("**/__*/*"):
-                dir.rmdir()
+                subprocess.run(["rm", "-rf", str(dir)])
```
```suggestion
                shutil.rmtree(dir)
```
This is cleaner.

sgugger replied:
It only works for dir, so you should make a test. It's still better than using `subprocess` for this.

mishig25 replied:
handled in https://github.com/huggingface/doc-builder/pull/139/commits/635938b0ec5030013266dc2d8299d4743056da52

sgugger commented on src/doc_builder/commands/preview.py:
```diff
@@ -175,9 +175,9 @@ def preview_command(args):
+                subprocess.run(["rm", "-rf", str(dir)])
             for f in output_path.glob("**/__*"):
-                f.unlink()
+                subprocess.run(["rm", "-rf", str(f)])
```
```suggestion
                os.unlink(f)
```

mishig25 commented:
@LysandreJik please feel free to retry the preview command.
You may need to reinstall the CLI if you face some error starting it up.

```
doc-builder preview datasets ~/Desktop/datasets/docs/source/
```

Fixed the relative path issue (should be working whether you provide absolute or relative path to docs folder)

LysandreJik commented:
Thanks for the ping, will try it again.
